pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,455 @@
+	}
+	p.Debug("releasing container %s from pool %s", c, pool)
+	podID := c.GetPodID()
+	pool.PodIDs[podID] = removeString(pool.PodIDs[podID], c.GetCacheID())
+	if len(pool.PodIDs[podID]) == 0 {
+		p.Debug("all containers removed, free pool allocation %s", pool.PrettyName())
+		// All containers of the pod removed from the pool.
+		// Free the pool for other pods.
+		delete(pool.PodIDs, podID)
+		pool.FreePodCapacity++
+	}
+	p.Debug("pool after release: %s", pool)
+	return nil
+}
+
+// UpdateResources is a resource allocation update request for this policy.
+func (p *podpools) UpdateResources(c cache.Container) error {
+	p.Debug("(not) updating container %s...", c.PrettyName())
+	return nil
+}
+
+// Rebalance tries to find an optimal allocation of resources for the current containers.
+func (p *podpools) Rebalance() (bool, error) {
+	p.Debug("(not) rebalancing containers...")
+	return false, nil
+}
+
+// HandleEvent handles policy-specific events.
+func (p *podpools) HandleEvent(*events.Policy) (bool, error) {
+	p.Debug("(not) handling event...")
+	return false, nil
+}
+
+// ExportResourceData provides resource data to export for the container.
+func (p *podpools) ExportResourceData(c cache.Container) map[string]string {
+	return nil
+}
+
+// Introspect provides data for external introspectiop.
+func (p *podpools) Introspect(*introspect.State) {
+	return
+}
+
+// allocatedPool returns a pool already allocated for the pod of a container
+func (p *podpools) allocatedPool(c cache.Container) *Pool {
+	podID := c.GetPodID()
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool { _, ok := pl.PodIDs[podID]; return ok })
+	if len(pools) == 0 {
+		return nil
+	}
+	return pools[0]
+}
+
+// allocatePool returns a pool allocated for the pod of a container
+func (p *podpools) allocatePool(c cache.Container) *Pool {
+	if pool := p.allocatedPool(c); pool != nil {
+		return pool
+	}
+	poolType, ok := p.getPoolType(c)
+	if !ok {
+		return nil
+	}
+	// Try to find a suitable pool and allocate it for the pod of
+	// the container.
+	podID := c.GetPodID()
+	// A pool must be of correct type.
+	pools := filterPools(p.pools,
+		func(pl *Pool) bool { return poolType.Name == pl.Type.Name })
+	// A pool must have capacity for at least one pod.
+	pools = filterPools(pools,
+		func(pl *Pool) bool { return pl.FreePodCapacity > 0 })
+	// Sort pools according to pool type fill order so that the
+	// first pool in the list is the preferred one.
+	switch poolType.FillOrder {
+	case FillBalanced:
+		sort.Slice(pools, func(i, j int) bool {
+			return pools[i].FreePodCapacity > pools[j].FreePodCapacity
+		})
+	case FillPacked:
+		sort.Slice(pools, func(i, j int) bool {
+			return pools[i].FreePodCapacity < pools[j].FreePodCapacity
+		})
+	case FillFirstFree:
+		// FirstFree is already the first of the pools list.
+	}
+	if len(pools) == 0 {
+		return nil
+	}
+	// Found a suitable pool. Allocate it for the pod.
+	pools[0].FreePodCapacity--
+	pools[0].PodIDs[podID] = []string{}
```

Comment:
These are very important remarks. Thanks, @klihub!

I'd suggest leaving 1 & 2 for future improvements (extended resources support) and addressing 3, 4 &5 in this patch.

Longer answers:
1. No plans for annotation-processing webhook. At the moment I'd prefer keeping the annotations as a "fully user-controlled method" of using podpools. It means that this would work always the same way, with or without a webhook in the cri-rm runtime configuration. This gives the user full power (and responsibility) to divide the CPU requests to containers as wanted. (CPU requests affect container process scheduling via cgroups `cpu.shares`.) To reduce user configuration burden (and power) in the future, the plan is to create extended resources from podpools on the nodes. Webhook-aided processing of CPU requests (in order to prevent overloading nodes) fits better in this mode of use, I think.
2. Proper resource accounting that supports multiple pool types will be enabled only with the extended resource implementation. Using a pooltype annotation requires user's actions to make sure the pod will be scheduled on a node with that pooltype being the only available pooltype. 
3. Check implemented.  Now it gives a warning on `overbooked cpuset`. Maybe later we can add configuration options on how to handle overbooking: refuse to run or something else. Or should we make refuse to run as a default behavior and allowing overbooking as an option?
4. Accounting and checking implemented: now reserved and shared CPUs are included in built-in pod pools and handled similarly as user-defined pool types. Except for that those types have unlimited pod capacity.
5. SetCPUShared implemented. I ended up setting it for all containers, even those in pod pools. I think that this policy should not prevent adjusting scheduling inside pod pools either.